shared/io/ktor/plugins/registry/PluginReference.kt

```diff
@@ -0,0 +1,226 @@
+    override fun toString() = "=="
+}
+
+/**
+ * Standard semantic version number references (i.e., 1.0.0)
+ */
+data class VersionNumber(
+    val number: String,
+    val mavenVersion: org.apache.maven.artifact.versioning.DefaultArtifactVersion = org.apache.maven.artifact.versioning.DefaultArtifactVersion(
+        number
+    )
+) : ArtifactVersion, org.apache.maven.artifact.versioning.ArtifactVersion by mavenVersion {
+    override fun contains(other: ArtifactVersion): Boolean = this == other
+    override fun asNumber(): VersionNumber = this
+    override fun resolve(version: VersionNumber): ArtifactVersion = version
+    override fun toString(): String = number
+}
+
+data class VersionRange(
+    private val range: org.apache.maven.artifact.versioning.VersionRange,
+    private val resolved: VersionNumber? = null,
+) : ArtifactVersion {
+    constructor(text: String): this(org.apache.maven.artifact.versioning.VersionRange.createFromVersionSpec(text))
+    override fun contains(other: ArtifactVersion): Boolean = other is VersionNumber && range.containsVersion(other.mavenVersion)
+    override fun resolve(version: VersionNumber): ArtifactVersion = copy(resolved = version)
+    override fun asRange(): VersionRange = this
+    override fun asNumber(): VersionNumber? = resolved
+    override fun toString(): String = range.toString()
+}
+
+/**
+ * Variable name for version represented in catalog (i.e., $ktor_version)
+ */
+data class VersionVariable(val name: String, val version: ArtifactVersion): ArtifactVersion by version {
+    override fun resolve(version: VersionNumber): ArtifactVersion =
+        VersionVariable(name, version)
+
+    override fun toString() = "\$$name@$version"
+}
+
+/**
+ * This will replace prefix versions ending with .+ with the corresponding Maven version range.
+ * For example, 2.+ -> [2,3)
+ */
+fun prefixVersionToMavenRange(text: String): String {
+    require(text.endsWith(".+")) { "Expected \"$text\" to end with .+" }
+    val prefix = text.substringBeforeLast(".+")
+    val nextPrefix = prefix.split('.').mapIndexed { index, part ->
+        if (index < prefix.count { it == '.' }) part else (part.toInt() + 1).toString()
+    }.joinToString(".")
+    return "[$prefix,$nextPrefix)"
+}
+
+class ArtifactReferenceSerializer: kotlinx.serialization.KSerializer<ArtifactReference> {
+    override val descriptor: kotlinx.serialization.descriptors.SerialDescriptor =
+        kotlinx.serialization.descriptors.PrimitiveSerialDescriptor(
+            "ArtifactReference",
+            kotlinx.serialization.descriptors.PrimitiveKind.STRING
+        )
+
+    override fun serialize(encoder: kotlinx.serialization.encoding.Encoder, value: ArtifactReference) {
+        encoder.encodeString(value.toString())
+    }
+
+    override fun deserialize(decoder: kotlinx.serialization.encoding.Decoder): ArtifactReference =
+        ArtifactReference.parse(decoder.decodeString())
+}
+
+// TODO maintain variables
```

Comment:
please log an issue for TODO

Reply:
Oh, I had done it, just forgot to remove the TODO.